typing/ctype.mli

```diff
@@ -441,12 +441,46 @@ val get_current_level: unit -> int
 val wrap_trace_gadt_instances: Env.t -> ('a -> 'b) -> 'a -> 'b
 val reset_reified_var_counter: unit -> unit
 
-val immediacy : Env.t -> type_expr -> Type_immediacy.t
-
 (* Stubs *)
 val package_subtype :
     (Env.t -> Path.t -> (Longident.t * type_expr) list ->
       Path.t -> (Longident.t * type_expr) list -> bool) ref
 
 (* Raises [Incompatible] *)
 val mcomp : Env.t -> type_expr -> type_expr -> unit
+
+(* [get_unboxed_type_representation] attempts to fully expand the input
+   type_expr, descending through [@@unboxed] types.  May fail in the case of
+   circular types or very deeply nested unboxed types, in which case it returns
+   the most expanded version it was able to compute.
+
+   [get_unboxed_type_approximation] does the same, but doesn't indicate whether
+   the type was fully expanded or not. *)
+val get_unboxed_type_representation :
+  Env.t -> type_expr -> (type_expr, type_expr) result
+val get_unboxed_type_approximation : Env.t -> type_expr -> type_expr
+
+(* [kind_immediacy_approx] may be a conservative approximation (return Unknown
+   for types that are actually immediate) in two cases: [@@unboxed] types, and
+   abbreviations (abstract types with a manifest).  *)
+val kind_immediacy_approx : type_decl_kind -> Type_immediacy.t
+val check_decl_immediate :
+  Env.t -> type_declaration -> Type_immediacy.t ->
+  (unit, Type_immediacy.Violation.t) result
+
+val check_type_immediate :
+  Env.t -> type_expr -> Type_immediacy.t ->
+  (unit, Type_immediacy.Violation.t) result
+
+type immediacy_context = Mode_crossing | Type_opt
+(* Bytecode is intended to be platform independent, so Immediate64 types are
+   treated as immediate only when targeting native code. We make an exception
+   for the mode crossing check, so that bytecode builds don't break locals code.
+   This means there's a potential unsoundness if you compile to bytecode on a
```

Comment:
[Looks like it](https://github.com/ocaml-flambda/ocaml-jst/blob/7e5a626e4b4e12f1e9106564e1baba4d0ef6309a/typing/typeopt.ml#L73)

Reply:
Looking more closely, it seems this fact is newly true in this patch. That is, before this patch, immediacy depends only on the target. That is, `immediate64` is immediate only when targetting 64-bit native code. (The written check you linked to checks the bit-width of the host system, but that width is always the same as the target.)

New in this patch, though, is to treat (for mode-crossing) `immediate64` as immediate when hosting on a 64-bit system, regardless of target (byte-code vs native). Instead, I think we want `immediate64` to count as immediate (for mode-crossing) when the target is 64-bit native, or when the target is byte-code. That is, if the target is byte-code, there should be no bit-width check at all. Then -- as long as we never enable stack allocation in a 32-bit byte-code environment -- we should be OK, and the bit-width of the host system does not affect the type system.

I'm worried I've missed something here, though, so push back if I've gotten this wrong.

Reply:
Almost - we do already treat imm64 differently for mode crossing (PR #97).  This PR expands on that slightly - see the end of the initial PR description.

Reply:
That is to say, there are two places where we use the `Mode_crossing` reason, and only the one in `Typecore.build_as_type_aux` is new.

Reply:
I think that explaining this in terms of the context of the check makes things more confusing that they need to be. Really there are two separate properties here:

1. Are values of this type always global?
2. Are values of this type always tagged integers (i.e. gc ignorable)?

For an `Immediate` type the answer to both of these is true. For an `Immediate64` type the answer to the first one is supposed to be true (although we haven't implemented the check for that properly yet, so it wouldn't be true on a 32bit system with locals enabled), whist the answer to the second one is true for native code on 64bit and false otherwise.

I think you should split off an `is_always_global` function in `ctype.ml` and an `is_always_gc_ignorable` or `is_always_tagged_int` function and put it in `typeopt.ml`. The reason to move `is_always_gc_ignorable` is because it is platform dependent and so shouldn't be called outside of `typeopt.ml`.

Also, maybe we should also just prevent 32bit configurations from using locals in `./configure` for now?

Reply:
+1 to @lpw25's suggestions.

Reply:
@lpw25 Sure, that sounds reasonable. To check: are you proposing to drop the `Sys.word_size` check from `is_always_global`?

Reply:
Yeah. For now we'll enforce the it through a configure time check -- or just the fact that we know not to use that configuration. Soon we'll be able to enforce it properly in the type system.

Reply:
OK, I've a commit that splits this into two functions and adds the configure-time check.

Please also be sure to have a look at the change in behavior in `Typecore.build_as_type_aux` mentioned in the initial PR description, and see if that looks OK.